File: src/main/java/com/paragon/client/systems/ui/animation/Easing.java
package com.paragon.client.systems.ui.animation;

import java.util.function.Function;

/**
 * Calculations from easings.net
 * @author Wolfsurge
 * @since 05/19/22
 */
public enum Easing {

    /**
     * No easing
     */
    LINEAR((input) -> input),

    /**
     * Speed gradually increases
     */
    CUBIC_IN((input) -> input * input * input),

    /**
     * Speed gradually decreases
     */
    CUBIC_OUT((input) -> 1 - Math.pow(1 - input, 3)),

    /**
     * Speed gradually increases until halfway and then decreases
     */
    CUBIC_IN_OUT((input) -> input < 0.5 ? 4 * input * input * input : 1 - Math.pow(-2 * input + 2, 3) / 2),

    /**
     * Speed gradually increases
     */
    QUINT_IN((input) -> input * input * input * input * input),

    /**
     * Speed gradually decreases
     */
    QUINT_OUT((input) -> 1 - Math.pow(1 - input, 5)),

    /**
     * Speed gradually increases until halfway and then decreases
     */
    QUINT_IN_OUT((input) -> input < 0.5 ? 16 * input * input * input * input * input : 1 - Math.pow(-2 * input + 2, 5) / 2),

    /**
     * Speed gradually increases
     */
    EXPO_IN((input) -> 4 * input * input * input),

    /**
     * Speed gradually decreases
     */
    EXPO_OUT((input) -> 1 - Math.pow(-2 * input + 2, 3) / 2),

    /**
     * Speed gradually increases until halfway and then decreases
     */
    EXPO_IN_OUT((input) -> input < 0.5 ? 4 * input * input * input : 1 - Math.pow(-2 * input + 2, 3) / 2);

    // The function that calculates the easing
    private final Function<Double, Double> easeFunction;

    Easing(Function<Double, Double> easeFunction) {
        this.easeFunction = easeFunction;
    }

    /**
     * Apply the easing function to the input
     * @param input The linear animation that we want to ease
     * @return The eased animation
     */
    public double ease(double input) {
        return easeFunction.apply(input);
    }

}

File: src/main/java/com/paragon/client/ui/util/animation/Easing.kt
package com.paragon.client.ui.util.animation

import kotlin.math.pow

/**
 * @author Surge
 */
@Suppress("unused")
enum class Easing(private val easeFunction: (Double) -> Double) {

    /**
     * No easing
     */
    LINEAR({ input -> input }),

    /**
     * Speed gradually increases
     */
    CUBIC_IN({ it * it * it }),

    /**
     * Speed gradually decreases
     */
    CUBIC_OUT({ 1 - (1 - it).pow(3.0) }),

    /**
     * Speed gradually increases until halfway and then decreases
     */
    CUBIC_IN_OUT({ if (it < 0.5) 4 * it * it * it else 1 - (-2 * it + 2).pow(3.0) / 2 }),

    /**
     * Speed gradually increases
     */
    QUINT_IN({ it * it * it * it * it }),

    /**
     * Speed gradually decreases
     */
    QUINT_OUT({ 1 - (1 - it).pow(5.0) }),

    /**
     * Speed gradually increases until halfway and then decreases
     */
    QUINT_IN_OUT({ if (it < 0.5) 16 * it * it * it * it * it else 1 - (-2 * it + 2).pow(5.0) / 2 }),

    /**
     * Speed gradually increases
     */
    EXPO_IN({ if (it == 0.0) 0.0 else 2.0.pow(10.0 * it - 10.0) }),

    /**
     * Speed gradually decreases
     */
    EXPO_OUT({ if (it == 1.0) 1.0 else 1 - 2.0.pow(-10 * it) }),

    /**
     * Speed gradually increases until halfway and then decreases
     */
    EXPO_IN_OUT({ if (it == 0.0) 0.0 else if (it == 1.0) 1.0 else if (it < 0.5) 2.0.pow(20 * it - 10) / 2.0 else (2 - 2.0.pow(-20 * it + 10)) / 2.0 });

    /**
     * Apply the easing function to the input
     *
     * @param input The linear animation that we want to ease
     * @return The eased animation
     */
    open fun ease(input: Double) = easeFunction.invoke(input)

}